seqr/utils/vcf_utils.py

```diff
@@ -0,0 +1,80 @@
+import re
+
+from collections import defaultdict
+
+from seqr.utils.middleware import ErrorsWarningsException
+from seqr.utils.file_utils import file_iter
+
+BLOCK_SIZE = 65536
+
+EXPECTED_META_FIELDS ={
+    'INFO': {
+        'AC': 'Integer',
+        'AN': 'Integer',
+        'AF': 'Float'
+    },
+    'FORMAT': {
+        'AD': 'Integer',
+        'DP': 'Integer',
+        'GQ': 'Integer',
+        'GT': 'String'
+    }
+}
+
+REQUIRED_HEADERS = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT']
+
+
+def _validate_vcf_header(header):
+    missing_headers = [h for h in REQUIRED_HEADERS if h not in header]
+    errors = []
+    if missing_headers:
+        missing_fields = ', '.join(missing_headers)
+        errors = [f'Missing required VCF header field(s) {missing_fields}.']
+    return errors
+
+
+def _validate_vcf_meta(meta):
+    errors = []
+    for field, sub_field_meta in EXPECTED_META_FIELDS.items():
+        missing_meta = [m for m in EXPECTED_META_FIELDS[field] if not meta.get(field, {}).get(m)]
+        if missing_meta:
+            missing_fields = ', '.join(missing_meta)
+            errors.append(f'Missing required {field} field(s) {missing_fields}')
+        for sub_field, expected in sub_field_meta.items():
+            value = meta.get(field, {}).get(sub_field)
+            if value and value != expected:
+                errors.append(f'Incorrect meta Type for {field}.{sub_field} - expected "{expected}", got "{value}"')
+    return errors
+
+
+def _get_vcf_meta_data(line):
+    r = re.search(r'##(?P<field>.*?)=<ID=(?P<id>[^,]*).*Type=(?P<type>[^,]*).*>$', line)
+    if r:
+        matches = r.groupdict()
+        return matches['field'], {matches['id']: matches['type']}
+    return None, None
+
+
+def validate_vcf_and_get_samples(vcf_filename):
+    byte_range = None if vcf_filename.endswith('.vcf') else (0, BLOCK_SIZE)
+    samples = set()
+    header = []
+    meta = defaultdict(dict)
+    for line in file_iter(vcf_filename, byte_range=byte_range):
+        if line.startswith('#'):
+            if line.startswith('#CHROM'):
+                header = line.rstrip().split('\t')
+                samples = set(header[len(REQUIRED_HEADERS):])
+            else:
+                field, sub_field_type = _get_vcf_meta_data(line)
+                if field:
+                    meta[field].update(sub_field_type)
+        else:
+            break
+
+    errors = _validate_vcf_header(header[0:len(REQUIRED_HEADERS)])
+    errors += _validate_vcf_meta(meta)
```

Comment:
sorry I was unclear her, but if `_validate_vcf_header` fails, it should hard fail and not continue with validating the meta. Each of these functions should raise an exception if there are any error

Reply:
In a VCF file, the meta info part comes first and doesn't depend on the header. So the meta info and the header validations are paralleled. It should be okay to provide all the errors at once.

Reply:
I just think its a different level of broken. If the header itself is wrong, the file is just completely useless, there is just no way you have a joint called VCF with any of those fields missing. If the meta info is wrong you probably do have a real VCF but maybe with some formatting issues. If you have a completely broken file, all the meta errors are just confusing to the user

Reply:
Okay, I'll raise the exceptions in the validation functions.